### PR TITLE
refactor: improve code quality

### PR DESCRIPTION
**Pull request by LomyW**
https://github.com/corona-school/project-user/issues/792

to do:
- [x] move logic to common
- [ ] generalize createZoomMeetingForAppointments
- [ ] check notification student_add_appointment_(group|match)`
- [x] mark all obsolete APIs with `@Deprecated`
- [x] restructure zoom folder
- [x] ensure every larger function is reached by at least one integration test
- [x] add feature toggle to chat (=> assert `TALKJS_APP_ID` and `TALKJS_API_KEY`)
- [x] fix conversation logic - ONHOLD until conversation is merged?
- [x] add assertion to `getConversationId`

### CODE REVIEW

File: common/appointment/create.ts [hunk 1]
```diff
@@ -115,10 +128,20 @@ export const createGroupAppointments = async (subcourseId: number, appointmentsT
                 })
         )
     );
+
+    // * send notification
+    for await (const participant of participants) {
```
Comment (Jonasdoubleyou) [inline common/appointment/create.ts:133]: No need for `for await`, `participants` is not an async iterator ...

File: common/appointment/create.ts [hunk 2]
```diff
@@ -150,3 +173,14 @@ export const createZoomMeetingForAppointments = async (
...
+
+export const saveZoomMeetingReport = async (appointment: Lecture) => {
+    const meetingReports: Prisma.JsonValue[] = appointment.zoomMeetingReport || [];
+    const result = await getZoomMeetingReport(appointment.zoomMeetingId);
+    meetingReports.push(result);
+
+    await prisma.lecture.update({
+        where: { id: appointment.id },
+        data: { zoomMeetingReport: meetingReports },
+    });
```
Comment (Jonasdoubleyou) [inline common/appointment/create.ts:185]: logging?

File: common/appointment/create.ts [hunk 3]
```diff
@@ -150,3 +173,14 @@ export const createZoomMeetingForAppointments = async (
         throw new Error(`Zoom - Error while creating zoom meeting: ${e}`);
     }
 };
+
+export const saveZoomMeetingReport = async (appointment: Lecture) => {
+    const meetingReports: Prisma.JsonValue[] = appointment.zoomMeetingReport || [];
```
Comment (Jonasdoubleyou) [inline common/appointment/create.ts:178]: So if two users do this at the same time, we lose one of the meeting reports?

Comment (Jonasdoubleyou) [reply common/appointment/create.ts:178]: Can't we `push` (https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-scalar-lists-arrays#adding-items-to-a-scalar-list)?

File: common/chat/create.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    } else if (reason === ContactReason.PARTICIPANT) {
+        await updateParticipantConversation(conversationId, subcourseId, conversation);
+    } else if (reason === ContactReason.PROSPECT) {
+        await updateProspectConversation(conversationId, subcourseId, conversation);
+    } else if (reason === ContactReason.CONTACT) {
+        await updateContactConversation(conversationId, conversationInfos);
+    }
+}
+
+async function updateMatchConversation(conversationId: string, match: any): Promise<void> {
```
Comment (Jonasdoubleyou) [inline common/chat/create.ts:72]: any?

File: common/chat/create.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    await updateConversation(updatedConversation);
+}
+
+async function updateParticipantConversation(conversationId: string, subcourseId: number | undefined, conversation: TJConversation): Promise<void> {
+    const subcoursesFromConversation = conversation?.custom.subcourse;
+    let subcourseIds = [];
+
+    try {
+        subcourseIds = JSON.parse(subcoursesFromConversation || '[]');
+    } catch (error) {}
```
Comment (Jonasdoubleyou) [inline common/chat/create.ts:89]: In which case would JSON.parse throw an error, and why would we silently ignore that?

File: common/chat/create.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    await updateConversation(updatedConversation);
+}
+
+async function updateProspectConversation(conversationId: string, subcourseId: number | undefined, conversation: TJConversation): Promise<void> {
+    const prospectSubcoursesFromConversation = conversation?.custom.prospectSubcourse;
+    let prospectSubcourseIds = [];
+
+    try {
+        prospectSubcourseIds = JSON.parse(prospectSubcoursesFromConversation || '[]');
+    } catch (error) {}
```
Comment (Jonasdoubleyou) [inline common/chat/create.ts:110]: same here

File: common/chat/create.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+    const conversationInfos: ConversationInfos = {
+        custom: {
+            ...(contact.match && { match: { matchId: contact.match.matchId } }),
+            ...(contact.subcourse && { subcourse: [...new Set(contact.subcourse)] }),
+        },
+    };
+
+    const conversation = await getOrCreateOneOnOneConversation([meUser, contactUser], conversationInfos, ContactReason.CONTACT);
+    return conversation.id;
```
Comment (Jonasdoubleyou) [inline common/chat/create.ts:177]: Generally we could use some logging here.

Comment (Jonasdoubleyou) [reply common/chat/create.ts:177]: Once we rollout the Chat you have to troubleshoot this, not sure if that is possible without logs ...

File: graphql/appointment/mutations.ts [hunk 8]
```diff
@@ -164,14 +164,7 @@ export class MutateAppointmentResolver {
...
         await hasAccess(context, 'Lecture', appointment);
-        const meetingReports: Prisma.JsonValue[] = appointment.zoomMeetingReport || [];
-        const result = await getZoomMeetingReport(appointment.zoomMeetingId);
-        meetingReports.push(result);
-
-        await prisma.lecture.update({
-            where: { id: appointmentId },
-            data: { zoomMeetingReport: meetingReports },
-        });
+        await saveZoomMeetingReport(appointment as Lecture);
```
Comment (Jonasdoubleyou) [inline graphql/appointment/mutations.ts:167]: Why is the typecast needed?

File: integration-tests/index.ts [hunk 9]
```diff
@@ -16,8 +16,9 @@ import './certificate';
 
 import './auth';
 import './settings';
-import './course';
 import './appointments';
```
Comment (Jonasdoubleyou) [inline integration-tests/index.ts:19]: The appointments test case depends on course, thus it deadlocks when appointments is run before course.

Comment (Jonasdoubleyou) [reply integration-tests/index.ts]: fixed it :)

**Review by Jonasdoubleyou — APPROVED** [review]

File: common/appointment/create.ts [hunk 10]
```diff
@@ -149,3 +172,16 @@ export const createZoomMeetingForAppointments = async (
...
```
Comment (Jonasdoubleyou) [inline common/appointment/create.ts:177]: Variables not needed?

**Review by Jonasdoubleyou — APPROVED** [review]

File: common/appointment/create.ts (hunk 10 above)
Comment (LomyW) [reply common/appointment/create.ts:177]: which variables do you mean?

Comment (LomyW) [reply common/appointment/create.ts:177]: which variables do you mean? @Jonasdoubleyou

File: common/appointment/create.ts [hunk 11]
```diff
@@ -153,6 +173,18 @@ export const createZoomMeetingForAppointments = async (
...
 };
 
+export const saveZoomMeetingReport = async (appointment: Lecture) => {
+    const meetingReports: Prisma.JsonValue[] = appointment.zoomMeetingReport || [];
+    const result = await getZoomMeetingReport(appointment.zoomMeetingId);
+    meetingReports.push(result);
+
+    await prisma.lecture.update({
+        where: { id: appointment.id },
+        data: { zoomMeetingReport: { push: meetingReports } },
```
Comment (Jonasdoubleyou) [inline common/appointment/create.ts:183]: Ah, actually push takes only the elements to add, not the whole array?

**Review by Jonasdoubleyou — CHANGES_REQUESTED** [review]

File: integration-tests/course.ts [hunk 12]
```diff
@@ -112,6 +116,64 @@ export const subcourseOne = test('Create Subcourse', async () => {
...
+                email: instructor.email,
+                display_name: instructor.firstname + ' ' + instructor.lastname,
+                personal_meeting_url: 'https://meet',
+            },
+        });
+
+        expectFetch({
+            url: 'https://api.zoom.us/v2/users/123/meetings',
+            method: 'POST',
+            body: '{"agenda":"My Meeting","default_password":false,"duration":60,"start_time":"*","timezone":"Europe/Berlin","type":2,"mute_upon_entry":true,"join_before_host":true,"waiting_room":true,"breakout_room":true,"settings":{"alternative_hosts":"","alternative_hosts_email_notification":false}}',
```
Comment (Jonasdoubleyou) [inline integration-tests/course.ts:151]: expects duration:60 here but adds duration: 15 below

**Review by Jonasdoubleyou — APPROVED** [review]

**Review by Jonasdoubleyou — APPROVED** [review]

**Pull request merged**